pkg/agent/cniserver/ipam/ipam_service.go

```diff
@@ -79,30 +89,52 @@ func ExecIPAMAdd(cniArgs *cnipb.CniCmdArgs, ipamType string, resultKey string) (
 	}
 
 	args := argsFromEnv(cniArgs)
-	driver := ipamDrivers[ipamType]
-	result, err := driver.Add(args, cniArgs.NetworkConfiguration)
-	if err != nil {
-		return nil, err
+
+	drivers := ipamDrivers[ipamType]
+	for _, driver := range drivers {
+		// Detect a driver that owns this request(f.e. based on ipam annotation
+		// of namespace or pod that initiated the request
+		if driver.Owns(args, k8sArgs, cniArgs.NetworkConfiguration) {
+			result, err := driver.Add(args, cniArgs.NetworkConfiguration)
+			if err != nil {
+				return nil, err
+			}
+			ipamResults.Store(resultKey, result)
+			return result, nil
+		}
 	}
-	ipamResults.Store(resultKey, result)
-	return result, nil
+
+	return nil, fmt.Errorf("No suitable IPAM driver found")
```

Comment:
We should never hit this however since a fallback driver should always "own" a request... is that correct?

Reply:
That depends on the driver - today, ipam_delegator owns everything, but not in (theoretical) general case.